### ISSUE TITLE
Add support for instrumenting a test with a decorator

### ISSUE THREAD

**Issue by alexamici** [nodev-io/pytest-nodev#19]
Enhancement request by @calmomau to support something along the lines of:
```python
@wish('parse_bool')
def test_parse_bool():
    assert not parse_bool('false')
    assert not parse_bool('0')
    assert parse_bool('true')
    assert parse_bool('1')
```
**Upadate:** Probably it can't be done see comments.

**Comment by alexamici** [comment]
Quite surprisingly in the simplest case the following decorator works:
```python
def wish(target):
    def wish_wrapper(test_func):
        def wish_wrapped(wish):
            globals()[target] = wish
            return test_func()
        return wish_wrapped
    return wish_wrapper
```
The general case is significantly more complex:
- targets like `mymodule.myfunc` are not in globals()
- it is not clear how to support `test_func` with fixtures other than wish

**Comment by alexamici** [comment]
The decorator in the previous comment only work is it is defined inside the test file that uses it due to the fact that we assign to globals() and that always refer to the namespace of the module where the decorator is defined, not the one where it is run.

**Comment by alexamici** [comment]
Upon further investigation the decorator doesn't look feasible. The test variables to be overridden are global variable, not free variables and overriding them with a ``wish`` for a test will interfere with all other decorated tests. :(

**Comment by alexamici** [comment]
The `monkeypatch` fixture come to the rescue:
```python
def wish(target_name):
    def wish_decorator(test_func):
        def wish_wrapper(wish, monkeypatch, *args, **kwargs):
            monkeypatch.setattr(target_name, wish)
            test_func(*args, **kwargs)
        return wish_wrapper
    return wish_decorator
```
It might be possible to use it. But apparently `target_name` must exist and must contain a dot `.`.

**Comment by alexamici** [comment]
This appears to work consistently:
```python
import inspect

def wish(target_name):
    def wish_decorator(test_func):
        def wish_wrapper(wish, monkeypatch, *args, **kwargs):
            if '.' in target_name:
                monkeypatch.setattr(target_name, wish)
            else:
                monkeypatch.setattr(inspect.getmodule(test_func), target_name, wish)
            return test_func(*args, **kwargs)
        return wish_wrapper
    return wish_decorator
```

**Comment by alexamici** [comment]
Closed with fd401107692921d1540722695067fe60550ec3a0.
